libyul/optimiser/VarNameCleaner.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/VarNameCleaner.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <algorithm>
+#include <cctype>
+#include <climits>
+#include <iterator>
+#include <string>
+#include <regex>
+
+using namespace yul;
+using namespace std;
+
+// ensure the following transforms
+// - input: a, a_1, a_1_2
+//	 output: a, a_1, a_2
+// - input: a, a_1, a_1_2, a_2
+//   output: a, a_1, a_2, a_3
+// - input: a_15, a_17
+//   output: a, a_2
+// - input: abi_decode_256
+//   output: abi_decode
+
+void VarNameCleaner::operator()(VariableDeclaration& _varDecl)
+{
+	for (TypedName& typedName: _varDecl.variables)
+		if (auto newName = makeCleanName(typedName.name.str()))
+			typedName.name = YulString{*newName};
+
+	ASTModifier::operator()(_varDecl);
+}
+
+void VarNameCleaner::operator()(Identifier& _identifier)
+{
+	if (auto newName = getCleanName(_identifier.name.str()))
+		_identifier.name = YulString{*newName};
+}
+
+boost::optional<string> VarNameCleaner::makeCleanName(string const& _name)
+{
+	if (auto newName = findCleanName(_name))
+	{
+		m_usedNames[*newName] = *newName;
```

Comment:
Does that really work if multiple variables are re-shuffled?

I think the module does not work properly for `{ let x_1 let x_2 let x_3 }`.

It first renames `x_1` to `x`, then `x_2` to `x_1`. With the second step, it will store `m_usedNames["x_1"] = "x_1"` and overwrite the renaming of `x_1` to `x`.

Can you add a test for that, please?

Reply:
I think this needs to be split into a set of used names in the translated source and a mapping of translations.

Reply:
I've added this test. thanks for it.

However, in your scenario, `x_1` gets assigned `x` (both names are remembered), `x_2` will indeed get its suffix stripped, and attempted to normalize it, but by coincidence, that's `x_2` again, as `x_1` is already remembered (as an "old variable name"). Merely the analog the same with `x_3`.

There is a way to split "newly used names" from "old introduced names". but I considered that not really worth enough (may change later?)

Reply:
Which test would that be? Maybe I'm blind, but I don't see it...

Reply:
Ah, now it's there :-).

Reply:
yes, I forgot to `git add` them before pushing up. :)